photosynthesis_metrics/fid.py

```diff
@@ -56,15 +51,99 @@ def __compute_fid(mu1: float, sigma1: np.ndarray, mu2: float, sigma2: np.ndarray
+        target_stack: activations from encoder with shape (N_samples, dims) and dtype: np.float32 in range 0 - 1
+    Returns:
+    --   : The Frechet Distance.
+    """
+    m_pred, s_pred = compute_statistics(predicted_stack)
+    m_targ, s_targ = compute_statistics(target_stack)
 
-    fid_gt_denoised = __compute_fid(m_gt, s_gt, m_denoised, s_denoised)
-    return fid_gt_denoised
+    fid_score = __compute_fid(m_pred, s_pred, m_targ, s_targ)
+    return fid_score
 
 
 def compute_statistics(stack: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
+    r"""Calculates the statistics used by FID
+    Args:
+        stack: activations from encoder with shape (N_samples, dims) and dtype: np.float32 in range 0 - 1
+    Returns:
+        mu:     mean over all activations from the encoder.
+        sigma:  covariance matrix over all activations from the encoder.
+    """
     mu = np.mean(stack, axis=0)
     sigma = np.cov(stack, rowvar=False)
     return mu, sigma
+
+class FID(nn.Module):
+    r"""Creates a criterion that measures Frechet Inception Distance score for two datasets of images
+    See https://arxiv.org/abs/1706.08500 for reference.
+    """
+
+    def __init__(self):
+        super(FID, self).__init__()
+
+
+    def forward(self, predicted_features: torch.Tensor, target_features: torch.Tensor) -> torch.Tensor:
+        r"""Interface of Frechet Inception Distance.
+        It's computed for a whole set of data and uses features from encoder instead of images itself to decrease computation cost.
+        FID can compare two data distributions with different number of samples. But dimensionalities should match, otherwise 
+        it won't be possible to correctly compute statistics.
+    
+        Args:
+            predicted_features: Low-dimension representation of predicted image set. Shape (N_pred, encoder_dim)
+            target_features: Low-dimension representation of target image set. Shape (N_targ, encoder_dim)
+            
+        Returns:
+            normed_msidx: normalized msid descriptor (if no `target_generator` specified).
+            msid_score: the scalar value of the distance between discriptor if both `prediction` and `target` given.
+        """
+
+        assert predicted_features.shape[1] == target_features.shape[1], \
+            f"Features dimensionalities should match, otherwise it won't be possible to correctly compute statistics. \
+                Got {predicted_features.shape[1]} and {target_features.shape[1]}"
+        assert predicted_features.dims == 2, f"Predicted features must have shape  (N_samples, encoder_dim), got {predicted_features.dims}"
+        assert target_features.dims == 2, f"Target features must have shape  (N_samples, encoder_dim), got {target_features.dims}"
+
+        score = compute_fid(predicted_features.numpy(), target_features.numpy())
+        return torch.Tensor(score)
```

Comment:
Please consider a conventional way to convert numpy arrays to torch tensors by using `torch.from_numpy()`. The difference is that `torch.Tensor()` always creates a float32 tensor, but `torch.from_numpy()` preserves the initial data type of the numpy array that was passed. 

For example:

```python
score = np.array([[1, 2, 3], [4, 5, 6]])
print(torch.from_numpy(score).dtype)
print(torch.Tensor(score).dtype)

>>> torch.int64
>>> torch.float32
```

Or

```python
score = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float64)
print(torch.from_numpy(score).dtype)
print(torch.Tensor(score).dtype)

>>> torch.float64
>>> torch.float32
```